File: 백준/Platinum/1435. 네 점/네 점.py
import sys
from math import sqrt
input = sys.stdin.readline


def area(a, b, c):
    s = a+b+c
    s /= 2
    k = s*(s-a)*(s-b)*(s-c)
    if k >= 0:
        return sqrt(s*(s-a)*(s-b)*(s-c))
    else:
        return -1

def isTriangle(a, b, c):
    l = sorted([a, b, c])
    if l[2] < l[0]+l[1]:
        return 1
    elif l[2] == l[0]+l[1]:
        return 0
    return -1


matrix = []
for i in range(4):
    matrix.append(list(map(int, input().split())))


def solve():
    mask = [(0, 1, 2, 3), (0, 1, 3, 2)]
    flag = 1

    i, j, k, e = (0, 1, 2, 3)
    a, b, c = matrix[i][j], matrix[j][k], matrix[k][i]
    t = isTriangle(a, b, c)
    if t == -1:
        return False
    x1 = (a ** 2 + b ** 2 - c ** 2) / (a << 1)
    y1 = sqrt(b ** 2 - x1 ** 2)

    i, j, k, e = (0, 1, 3, 2)
    a, b, c = matrix[i][j], matrix[j][k], matrix[k][i]
    t = isTriangle(a, b, c)
    if t == -1:
        return False
    x2 = (a ** 2 + b ** 2 - c ** 2) / (a << 1)
    y2 = sqrt(b ** 2 - x2 ** 2)
    if sqrt((x2-x1)**2 + (y2-y1)**2) <= matrix[2][3] <= sqrt((x2-x1)**2 + (y2+y1)**2):
        return True
    return False



if solve():
    print("YES")
else:
    print("NO")

File: 백준/Platinum/1435. 네 점/네 점.cc
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;


int isTriangle(long double a, long double b, long double c) {
    vector<long double> l = {a, b, c};
    sort(l.begin(), l.end());
    if (l[2] < l[0] + l[1]) {
        return 1; 
    } else if (l[2] == l[0] + l[1]) {
        return 0; 
    }
    return -1; 
}


bool solve(const vector<vector<long double>>& matrix) {
    long double i, j, k;
    long double a, b, c;
    long double x1, y1, x2, y2;

    
    i = 0; j = 1; k = 2;
    a = matrix[i][j]; b = matrix[k][i]; c = matrix[j][k];
    int t = isTriangle(a, b, c);
    if (t == -1) return false;
    x1 = (a * a + b * b - c * c) / (2 * a);
    y1 = sqrt(b * b - x1 * x1);
    

    
    i = 0; j = 1; k = 3;
    a = matrix[i][j]; b = matrix[k][i]; c = matrix[j][k];
    t = isTriangle(a, b, c);
    if (t == -1) return false;
    x2 = (a * a + b * b - c * c) / (2 * a);
    y2 = sqrt(b * b - x2 * x2);
    

    
    long double dist1 = sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));
    long double dist2 = sqrt((x2 - x1) * (x2 - x1) + (y2 + y1) * (y2 + y1));
    return dist1 <= matrix[2][3] && matrix[2][3] <= dist2;
}

int main() {
    
    vector<vector<long double>> matrix(4, vector<long double>(4));
    for (int i = 0; i < 4; ++i) {
        for (int j = 0; j < 4; ++j) {
            cin >> matrix[i][j];
        }
    }

  
    if (solve(matrix)) {
        cout << "YES" << endl;
    } else {
        cout << "NO" << endl;
    }

    return 0;
}
